elbox_WRGND15/progs.gpl/inotify/inotify-tools-3.13/man/inotifywait.1
.TH inotifywait 1 "January 01, 2008" "inotifywait 3.13"

.SH NAME
inotifywait \- wait for changes to files using inotify

.SH SYNOPSIS
.B inotifywait
.RB [ \-hcmrq ]
.RB [ \-e
<event> ]
.RB [ \-t
<seconds> ]
.RB [ \-\-format
<fmt> ]
.RB [ \-\-timefmt
<fmt> ]
<file> [ ... ]

.SH DESCRIPTION
.B inotifywait
efficiently waits for changes to files using Linux's
.BR inotify(7)
interface.  It is suitable for waiting for changes to files from shell scripts.
It can either exit once an event occurs, or continually execute and output events
as they occur.

.SH OUTPUT
.B inotifywait
will output diagnostic information on standard error and event information on
standard output.  The event output can be configured, but by default it
consists of lines of the following form:

.B watched_filename EVENT_NAMES event_filename

.TP
.B watched_filename
is the name of the file on which the event occurred.  If the file is a directory,
a trailing slash is output.
.TP
.B EVENT_NAMES
are the names of the inotify events which occurred, separated by commas.
.TP
.B event_filename
is output only when the event occurred on a directory, and in this case the name
of the file within the directory which caused this event is output.

By default, any special characters in filenames are not escaped in any way.  This
can make the output of inotifywait difficult to parse in awk scripts or similar.
The
.B \-\-csv
and
.B \-\-format
options will be helpful in this case.

.SH OPTIONS
.TP
.B \-h, \-\-help
Output some helpful usage information.
.TP
.B @<file>
When watching a directory tree recursively, exclude the specified file from
being watched.  The file must be specified with a relative or absolute path
according to whether a relative or absolute path is given for watched
directories.  If a specific path is explicitly both included and excluded, it
will always be watched.

.B Note:
If you need to watch a directory or file whose name starts with @, give the
absolute path.
.TP
.B \-\-fromfile <file>
Read filenames to watch or exclude from a file, one filename per line.  If
filenames begin with @ they are excluded as described above.  If <file> is `-',
filenames are read from standard input.  Use this option if you need to watch
too many files to pass in as command line arguments.
.TP
.B \-m, \-\-monitor
Instead of exiting after receiving a single event, execute indefinitely.  The
default behaviour is to exit after the first event occurs.
.TP
.B \-r, \-\-recursive
Watch all subdirectories of any directories passed as arguments.  Watches
will be set up recursively to an unlimited depth.  Symbolic links are not
traversed.  Newly created subdirectories will also be watched.

.B Warning:
If you use this option while watching the root directory
of a large tree, it may take quite a while until all inotify watches are
established, and events will not be received in this time.  Also, since one
inotify watch will be established per subdirectory, it is possible that the
maximum amount of inotify watches per user will be reached.  The default
maximum is 8192; it can be increased by writing to
.BR /proc/sys/fs/inotify/max_user_watches .

.TP
.B \-q, \-\-quiet
If specified once, the program will be less verbose.  Specifically, it will not
state when it has completed establishing all inotify watches.

If specified twice, the program will output nothing at all, except in the case
of fatal errors.

.TP
.B \-\-exclude <pattern>
Do not process any events whose filename matches the specified POSIX extended
regular expression, case sensitive.

.TP
.B \-\-excludei <pattern>
Do not process any events whose filename matches the specified POSIX extended
regular expression, case insensitive.

.TP
.B \-t <seconds>, \-\-timeout <seconds>
Listen for an event for the specified amount of seconds, exiting if an event
has not occurred in that time.

.TP
.B \-e <event>, \-\-event <event>
Listen for specific event(s) only.  The events which can be listened for are
listed in the
.B EVENTS
section.  This option can be specified more than once.  If omitted, all events
are listened for.

.TP
.B \-c, \-\-csv
Output in CSV (comma-separated values) format.  This is useful when filenames
may contain spaces, since in this case it is not safe to simply split the output
at each space character.

.TP
.B \-\-timefmt <fmt>
Set a time format string as accepted by strftime(3) for use with the `%T' conversion
in the \-\-format option.

.TP
.B \-\-format <fmt>
Output in a user-specified format, using printf-like syntax.  The event strings
output are limited to around 4000 characters and will be truncated to this length.
The following conversions are supported:

.TP
%w
This will be replaced with the name of the Watched file on which an event occurred.

.TP
%f
When an event occurs within a directory, this will be replaced with the name of the
File which caused the event to occur.  Otherwise, this will be replaced with an
empty string.

.TP
%e
Replaced with the Event(s) which occurred, comma-separated.

.TP
%Xe
Replaced with the Event(s) which occurred, separated by whichever character is
in the place of `X'.

.TP
%T
Replaced with the current Time in the format specified by the \-\-timefmt option,
which should be a format string suitable for passing to strftime(3).



.SH "EXIT STATUS"
.TP
.B 0
The program executed successfully, and an event occurred which was being
listened for.
.TP
.B 1
An error occurred in execution of the program, or an event occurred which was
not being listened for.  The latter generally occurs if something happens which
forcibly removes the inotify watch, such as a watched file being deleted or the
filesystem containing a watched file being unmounted.
.TP
.B 2
The
.B \-t
option was used and an event did not occur in the specified interval of time.

.SH EVENTS
The following events are valid for use with the
.B \-e
option:

.TP
.B access
A watched file or a file within a watched directory was read from.

.TP
.B modify
A watched file or a file within a watched directory was written to.

.TP
.B attrib
The metadata of a watched file or a file within a watched directory was
modified.  This includes timestamps, file permissions, extended attributes etc.

.TP
.B close_write
A watched file or a file within a watched directory was closed, after being
opened in writeable mode.  This does not necessarily imply the file was written
to.

.TP
.B close_nowrite
A watched file or a file within a watched directory was closed, after being
opened in read-only mode.

.TP
.B close
A watched file or a file within a watched directory was closed, regardless of
how it was opened.  Note that this is actually implemented simply by listening
for both
.B close_write
and
.B close_nowrite,
hence all close events received will be output as one of these, not
.B CLOSE.

.TP
.B open
A watched file or a file within a watched directory was opened.

.TP
.B moved_to
A file or directory was moved into a watched directory.  This event occurs even
if the file is simply moved from and to the same directory.

.TP
.B moved_from
A file or directory was moved from a watched directory.  This event occurs even
if the file is simply moved from and to the same directory.

.TP
.B move
A file or directory was moved from or to a watched directory.  Note that this is
actually implemented simply by listening for both
.B moved_to
and
.B moved_from,
hence all close events received will be output as one or both of these, not
.B MOVE.

.TP
.B create
A file or directory was created within a watched directory.

.TP
.B delete
A file or directory within a watched directory was deleted.

.TP
.B delete_self
A watched file or directory was deleted.  After this event the file or directory
is no longer being watched.  Note that this event can occur even if it is not
explicitly being listened for.

.TP
.B unmount
The filesystem on which a watched file or directory resides was unmounted.
After this event the file or directory is no longer being watched.  Note that
this event can occur even if it is not explicitly being listened to.


.SH EXAMPLES

.SS Example 1
Running inotifywait at the command-line to wait for any file in the `test'
directory to be accessed.  After running inotifywait, `cat test/foo' is run
in a separate console.

.nf
% inotifywait test
Setting up watches.
Watches established.
test/ ACCESS foo
.fi

.SS Example 2
A short shell script to efficiently wait for httpd-related log messages and
do something appropriate.

.nf
#!/bin/sh
while inotifywait -e modify /var/log/messages; do
  if tail -n1 /var/log/messages | grep httpd; then
    kdialog --msgbox "Apache needs love!"
  fi
done
.fi

.SS Example 3
A custom output format is used to watch `~/test'.  Meanwhile, someone runs
`touch ~/test/badfile; touch ~/test/goodfile; rm ~/test/badfile' in another
console.

.nf
% inotifywait -m -r --format '%:e %f' ~/test
Setting up watches.  Beware: since -r was given, this may take a while!
Watches established.
CREATE badfile
OPEN badfile
ATTRIB badfile
CLOSE_WRITE:CLOSE badfile
CREATE goodfile
OPEN goodfile
ATTRIB goodfile
CLOSE_WRITE:CLOSE goodfile
DELETE badfile
.fi


.SH BUGS
There are race conditions in the recursive directory watching code
which can cause events to be missed if they occur in a directory immediately
after that directory is created.  This is probably not fixable.

It is assumed the inotify event queue will never overflow.

.SH AUTHORS
inotifywait is written and maintained by Rohan McGovern <rohan@mcgovern.id.au>.

inotifywait is part of inotify-tools.  The inotify-tools website is located at:
.I http://inotify-tools.sourceforge.net/

.SH "SEE ALSO"
inotifywatch(1), strftime(3), inotify(7)


elbox_WRGND15/progs.gpl/inotify/inotify-tools-3.13/src/inotifywait.c
// kate: replace-tabs off; space-indent off;

#include "../config.h"
#include "common.h"

#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <fcntl.h>
#include <errno.h>
#include <assert.h>
#include <regex.h>

#include <inotifytools/inotifytools.h>
#include <inotifytools/inotify.h>

#include <getopt.h>

extern char *optarg;
extern int optind, opterr, optopt;

char * MediaServerComm = "M"; //default M: uPNP, I: iTune

#define EXIT_OK 0
#define EXIT_ERROR 1
#define EXIT_TIMEOUT 2

#define MAX_STRLEN 4096
#define EXCLUDE_CHUNK 1024

#define nasprintf(...) niceassert( -1 != asprintf(__VA_ARGS__), "out of memory")

#define UPDATE_ADD				1
#define UPDATE_DELETE			2
#define UPDATE_DELETE_DIR	3
#define UPDATE_RENAME			4

//ino_t getID(char *path);
bool UpdateDB(int updateFlag, char *FilePath1, char *FilePath2,char *cb_script);

// METHODS
bool parse_opts(
  int * argc,
  char *** argv,
  int * events,
  bool * monitor,
  int * quiet,
  unsigned long int * timeout,
  int * recursive,
  bool * csv,
  char ** format,
  char ** timefmt,
  char ** fromfile,
  char ** regex,
  char ** iregex,
  char **cb_script
);

void print_help();

char * csv_escape( char * string ) {
	static char csv[MAX_STRLEN+1];
	static unsigned int i, ind;

	if ( strlen(string) > MAX_STRLEN ) {
		return NULL;
	}

	if ( strlen(string) == 0 ) {
		return NULL;
	}

	// May not need escaping
	if ( !strchr(string, '"') && !strchr(string, ',') && !strchr(string, '\n')
	     && string[0] != ' ' && string[strlen(string)-1] != ' ' ) {
		strcpy( csv, string );
		return csv;
	}

	// OK, so now we _do_ need escaping.
	csv[0] = '"';
	ind = 1;
	for ( i = 0; i < strlen(string); ++i ) {
		if ( string[i] == '"' ) {
			csv[ind++] = '"';
		}
		csv[ind++] = string[i];
	}
	csv[ind++] = '"';
	csv[ind] = '\0';

	return csv;
}



void validate_format( char * fmt ) {
	// Make a fake event
	struct inotify_event * event =
	   (struct inotify_event *)malloc(sizeof(struct inotify_event) + 4);
	if ( !event ) {
		fprintf( stderr, "Seem to be out of memory... yikes!\n" );
		exit(EXIT_ERROR);
	}
	event->wd = 0;
	event->mask = IN_ALL_EVENTS;
	event->len = 3;
	strcpy( event->name, "foo" );
	FILE * devnull = fopen( "/dev/null", "a" );
	if ( !devnull ) {
		fprintf( stderr, "Couldn't open /dev/null: %s\n", strerror(errno) );
		free( event );
		return;
	}
	if ( -1 == inotifytools_fprintf( devnull, event, fmt ) ) {
		fprintf( stderr, "Something is wrong with your format string.\n" );
		exit(EXIT_ERROR);
	}
	free( event );
	fclose(devnull);
}


void output_event_csv( struct inotify_event * event ) {
	printf("%s,", csv_escape( inotifytools_filename_from_wd( event->wd )));
	printf("%s,", csv_escape( inotifytools_event_to_str( event->mask ) ) );
	if ( event->len > 0 )
		printf("%s", csv_escape( event->name ) );
	printf("\n");
}

int quiet = 0;

#if 0 //Added by Ben 200.06.25
int Transfer_Path(char *read_path, char *full_path, int rootflag)
{
	char *p1;
	int id = atoi(read_path); 

	memset(full_path, 0x00, sizeof(full_path));
	
	if ((p1 = strchr(read_path, '/')) != NULL)		
	{
		*p1 = 0x00;
		p1++;
		if (id <= 100)
		{
			if (rootflag == 1)
				sprintf(full_path, "/mnt/vg%s/lvnas_default/", read_path);
			else
				sprintf(full_path, "/mnt/vg%s/lvnas_default/%s", read_path, p1);
		}
		else
		{
			if (rootflag == 1)
				sprintf(full_path, "/mnt/md%s/", read_path);
			else
				sprintf(full_path, "/mnt/md%s/%s", read_path, p1);
		}
		p1--;
		*p1 = '/';
	}
	else
	{
		if (id <= 100)
			sprintf(full_path, "/mnt/vg%s/lvnas_default", read_path);
		else
			sprintf(full_path, "/mnt/md%s", read_path);
	}
	return 0;
}
#endif

int main(int argc, char ** argv)
{
	int events = 0;
	bool monitor = false;
	
	unsigned long int timeout = 0;
	int recursive = 0;
	bool csv = false;
	char * format = NULL;
	char * timefmt = NULL;
	char * fromfile = NULL;
	char * regex = NULL;
	char * iregex = NULL;
	char *cb_script=NULL;//marco
	
	char myname[512];
	memset(myname, 0x00, sizeof(myname));
	strcpy(myname, argv[0]);
	
	// Parse commandline options, aborting if something goes wrong
	if ( !parse_opts(&argc, &argv, &events, &monitor, &quiet, &timeout,
	                 &recursive, &csv, &format, &timefmt, &fromfile,
	                 &regex, &iregex,&cb_script) ) {
		return EXIT_ERROR;
	}
	
#if 1 //Added by Ben 2008.04.03
	char* p =	strtok(myname, "/");
	do
	{
		if (strcmp(p, "inotify_uPNP") == 0 )
		{
			if (quiet == 0) printf("Server Type = uPNP\n");
			MediaServerComm = "M"; //UPNP Server
			break;
		}
		else if (strcmp(p, "inotify_iTune") == 0 )
		{
			if (quiet == 0) printf("Server Type = iTune\n");
			MediaServerComm = "I"; //iTune Server
			break;
		}
	}	while(p = strtok(NULL, "/"));
#endif

	if ( !inotifytools_initialize() )
	{
		fprintf(stderr, "Couldn't initialize inotify.  Are you running Linux "
		                "2.6.13 or later, and was the\n"
		                "CONFIG_INOTIFY option enabled when your kernel was "
		                "compiled?  If so, \n"
		                "something mysterious has gone wrong.  Please e-mail "
		                PACKAGE_BUGREPORT "\n"
		                " and mention that you saw this message.\n");
		return EXIT_ERROR;
	}

	if ( timefmt ) inotifytools_set_printf_timefmt( timefmt );
	if ((regex && !inotifytools_ignore_events_by_regex(regex, REG_EXTENDED) ) ||
		  (iregex && !inotifytools_ignore_events_by_regex(iregex, REG_EXTENDED|REG_ICASE)))
	{
		fprintf(stderr, "Error in `exclude' regular expression.\n");
		return EXIT_ERROR;
	}

	if ( format ) validate_format(format);

	// Attempt to watch file
	// If events is still 0, make it all events.
	if (events == 0)
		events = IN_ALL_EVENTS;

#if 1 //Added by Ben 2008.06.26
	char full_path[512];
	memset(full_path, 0x00, sizeof(full_path));
	if (strcmp(argv[0], "root") == 0)
		strcpy(full_path, "/mnt/");
	else
		strcpy(full_path,argv[0]);
//#ifndef DIR685NAS	
//		Transfer_Path(argv[0], full_path, 0);
//#endif		
	argv[0] = full_path;
#endif
	FileList list = construct_path_list( argc, argv, fromfile );

	if (0 == list.watch_files[0])
	{
		fprintf(stderr, "No files specified to watch!\n");
		return EXIT_ERROR;
	}

	if ( !quiet )
	{
		fprintf( stderr, "Setting up watches.  " );
		if ( recursive )
			fprintf( stderr, "Beware: since -r was given, this may take a while!" );
			
		fprintf( stderr, "\n" );
	}

	// now watch files
	for ( int i = 0; list.watch_files[i]; ++i )
	{
		char const *this_file = list.watch_files[i];
		if ( (recursive && !inotifytools_watch_recursively_with_exclude(
		                        this_file,
		                        events,
		                        list.exclude_files ))
		     || (!recursive && !inotifytools_watch_file( this_file, events )) )
		{
			if ( inotifytools_error() == ENOSPC ) 
			{
				fprintf(stderr, "Failed to watch %s; upper limit on inotify "
				                "watches reached!\n", this_file );
				fprintf(stderr, "Please increase the amount of inotify watches "
				        "allowed per user via `/proc/sys/fs/inotify/"
				        "max_user_watches'.\n");
			}
			else 
			{
				fprintf(stderr, "Couldn't watch %s: %s\n", this_file,
				        strerror( inotifytools_error() ) );
			}
			return EXIT_ERROR;
		}
	}

	if ( !quiet )
		fprintf( stderr, "Watches established.\n" );

	// Now wait till we get event
	struct inotify_event * event;
	char * moved_from = 0;
	
	do {
		event = inotifytools_next_event( timeout );

		if ( !event )
		{
			if ( !inotifytools_error() ) {
				return EXIT_TIMEOUT;
			}
			else {
				fprintf(stderr, "%s\n", strerror( inotifytools_error() ) );
				return EXIT_ERROR;
			}
		}

		if ( quiet < 2 )
		{
			if ( csv ) {
				output_event_csv( event );
			}
			else if ( format ) {
				inotifytools_printf( event, format );
			}
			else {
				;//inotifytools_printf( event, "Event:%e, filename = %w%f\n" );
			}
		}
		
		// if we last had MOVED_FROM and don't currently have MOVED_TO,
		// moved_from file must have been moved outside of tree - so unwatch it.
		if ( moved_from && !(event->mask & IN_MOVED_TO) )
		{
			if ( !inotifytools_remove_watch_by_filename( moved_from ) )
				fprintf( stderr, "Error removing watch on %s: %s\n", moved_from, strerror(inotifytools_error()) );
			free( moved_from );
			moved_from = 0;
		}

		if ( monitor && recursive )
		{
			if ((event->mask & IN_CREATE) || (!moved_from && (event->mask & IN_MOVED_TO)))
			{
				// New file - if it is a directory, watch it
				static char * new_file;

				nasprintf( &new_file, "%s%s", inotifytools_filename_from_wd( event->wd ), event->name );

				if ( isdir(new_file) && !inotifytools_watch_recursively( new_file, events ) )
					fprintf( stderr, "Couldn't watch new directory %s: %s\n", new_file, strerror( inotifytools_error() ) );
				
				UpdateDB(UPDATE_ADD, new_file, '\0',cb_script);
				free( new_file );
			} // IN_CREATE
			else if (event->mask & IN_MOVED_FROM)
			{
				// if not watched...
				if (event->mask & IN_ISDIR)
				{
					nasprintf( &moved_from, "%s%s/", inotifytools_filename_from_wd( event->wd ), event->name );
					if ( inotifytools_wd_from_filename(moved_from) == -1 )
					{
						free( moved_from );
						moved_from = 0;
					}
				}
				else
				{
					nasprintf( &moved_from, "%s%s", inotifytools_filename_from_wd( event->wd ), event->name );
				}
			} // IN_MOVED_FROM
			else if (event->mask & IN_MOVED_TO)
			{
				if ( moved_from )
				{
					static char * new_name;
					if (event->mask & IN_ISDIR)						
						nasprintf( &new_name, "%s%s/", inotifytools_filename_from_wd( event->wd ), event->name );
					else
						nasprintf( &new_name, "%s%s", inotifytools_filename_from_wd( event->wd ), event->name );
					
					inotifytools_replace_filename( moved_from, new_name );

					UpdateDB(UPDATE_RENAME, moved_from, new_name,cb_script);

					free( new_name );
					free( moved_from );
					moved_from = 0;
				} // moved_from
			}
			else if (event->mask & IN_DELETE)
			{
				static char * del_name;
				if (event->mask & IN_ISDIR )
				{
					nasprintf( &del_name, "%s%s/", inotifytools_filename_from_wd( event->wd ), event->name );
					UpdateDB(UPDATE_DELETE_DIR, del_name, '\0',cb_script);
				}
				else
				{
					nasprintf( &del_name, "%s%s", inotifytools_filename_from_wd( event->wd ), event->name );
					UpdateDB(UPDATE_DELETE, del_name, '\0',cb_script);
				}
				
				free( del_name );
			}
		}
		
		fflush( NULL );

	} while ( monitor );
	
//	if(cb_script)
//		free(cb_script);//marco
		
	// If we weren't trying to listen for this event...
	if ( (events & event->mask) == 0 )
		return EXIT_ERROR;

	return EXIT_OK;
}


bool UpdateDB(int updateFlag, char *FilePath1, char *FilePath2,char *cb_script)
{
	if (!updateFlag || !FilePath1 || !cb_script) return 0;

	int pid;
	char File[1024];
	pid = vfork();
	
	if (pid) 
	{
		waitpid(pid, NULL, 0);
		return 1;
	}
	else if (pid == 0) 
	{
		/* close fd's? */
		/* exec script */
		switch(updateFlag)
		{
			case UPDATE_ADD:
				sprintf(File,"%s%s",FilePath1, (isdir(FilePath1) ? "/" : "") );
				execlp(cb_script, cb_script, "ADD",File, NULL);
				exit(1);
				break;
			
			case UPDATE_DELETE:
				execlp(cb_script, cb_script, "DEL",FilePath1, NULL);
				exit(1);
				break;
		
			case UPDATE_DELETE_DIR:
				execlp(cb_script, cb_script, "DEL_DIR",FilePath1, NULL);
				exit(1);
				break;
			
			case UPDATE_RENAME:
				execlp(cb_script, cb_script, "RENAME",FilePath1,FilePath2, NULL);
				exit(1);
				break;
		}			
	}
	return 1;
}



bool parse_opts(
  int * argc,
  char *** argv,
  int * events,
  bool * monitor,
  int * quiet,
  unsigned long int * timeout,
  int * recursive,
  bool * csv,
  char ** format,
  char ** timefmt,
  char ** fromfile,
  char ** regex,
  char ** iregex,
  char **cb_script
) {
	assert( argc ); assert( argv ); assert( events ); assert( monitor );
	assert( quiet ); assert( timeout ); assert( csv ); assert( format );
	assert( timefmt ); assert( fromfile ); assert( regex ); assert( iregex );

	// Short options
	char * opt_string = "mrhcqt:fe:s:";

	// Construct array
	struct option long_opts[15];

	// --help
	long_opts[0].name = "help";
	long_opts[0].has_arg = 0;
	long_opts[0].flag = NULL;
	long_opts[0].val = (int)'h';
	// --event
	long_opts[1].name = "event";
	long_opts[1].has_arg = 1;
	long_opts[1].flag = NULL;
	long_opts[1].val = (int)'e';
	int new_event;
	// --monitor
	long_opts[2].name = "monitor";
	long_opts[2].has_arg = 0;
	long_opts[2].flag = NULL;
	long_opts[2].val = (int)'m';
	// --quiet
	long_opts[3].name = "quiet";
	long_opts[3].has_arg = 0;
	long_opts[3].flag = NULL;
	long_opts[3].val = (int)'q';
	// --timeout
	long_opts[4].name = "timeout";
	long_opts[4].has_arg = 1;
	long_opts[4].flag = NULL;
	long_opts[4].val = (int)'t';
	char * timeout_end = NULL;
	// --filename
	long_opts[5].name = "filename";
	long_opts[5].has_arg = 0;
	long_opts[5].flag = NULL;
	long_opts[5].val = (int)'f';
	// --recursive
	long_opts[6].name = "recursive";
	long_opts[6].has_arg = 0;
	long_opts[6].flag = NULL;
	long_opts[6].val = (int)'r';
	// --csv
	long_opts[7].name = "csv";
	long_opts[7].has_arg = 0;
	long_opts[7].flag = NULL;
	long_opts[7].val = (int)'c';
	// --format
	long_opts[8].name = "format";
	long_opts[8].has_arg = 1;
	long_opts[8].flag = NULL;
	long_opts[8].val = (int)'n';
	// format with trailing newline
	static char * newlineformat;
	// --timefmt
	long_opts[9].name = "timefmt";
	long_opts[9].has_arg = 1;
	long_opts[9].flag = NULL;
	long_opts[9].val = (int)'i';
	// --fromfile
	long_opts[10].name = "fromfile";
	long_opts[10].has_arg = 1;
	long_opts[10].flag = NULL;
	long_opts[10].val = (int)'z';
	// --exclude
	long_opts[11].name = "exclude";
	long_opts[11].has_arg = 1;
	long_opts[11].flag = NULL;
	long_opts[11].val = (int)'a';
	// --excludei
	long_opts[12].name = "excludei";
	long_opts[12].has_arg = 1;
	long_opts[12].flag = NULL;
	long_opts[12].val = (int)'b';
	//marco
	long_opts[13].name = "script";
	long_opts[13].has_arg =1;
	long_opts[13].flag = NULL;
	long_opts[13].val = (int)'s';
	
	long_opts[14].name = 0;
	long_opts[14].has_arg = 0;
	long_opts[14].flag = 0;
	long_opts[14].val = 0;

	// Get first option
	char curr_opt = getopt_long(*argc, *argv, opt_string, long_opts, NULL);

	// While more options exist...
	while ( (curr_opt != '?') && (curr_opt != (char)-1) )
	{
		switch ( curr_opt )
		{
			// --help or -h
			case 'h':
				print_help();
				// Shouldn't process any further...
				return false;
				break;

			// --monitor or -m
			case 'm':
				*monitor = true;
				break;

			// --quiet or -q
			case 'q':
				(*quiet)++;
				break;

			// --recursive or -r
			case 'r':
				(*recursive)++;
				break;

			// --csv or -c
			case 'c':
				(*csv) = true;
				break;

			// --filename or -f
			case 'f':
				fprintf(stderr, "The '--filename' option no longer exists.  "
				                "The option it enabled in earlier\nversions of "
				                "inotifywait is now turned on by default.\n");
				return false;
				break;

			// --format
			case 'n':
				newlineformat = (char *)malloc(strlen(optarg)+2);
				strcpy( newlineformat, optarg );
				strcat( newlineformat, "\n" );
				(*format) = newlineformat;
				break;

			// --timefmt
			case 'i':
				(*timefmt) = optarg;
				break;

			// --exclude
			case 'a':
				(*regex) = optarg;
				break;

			// --excludei
			case 'b':
				(*iregex) = optarg;
				break;

			// --fromfile
			case 'z':
				if (*fromfile) {
					fprintf(stderr, "Multiple --fromfile options given.\n");
					return false;
				}
				(*fromfile) = optarg;
				break;

			// --timeout or -t
			case 't':
				*timeout = strtoul(optarg, &timeout_end, 10);
				if ( *timeout_end != '\0' )
				{
					fprintf(stderr, "'%s' is not a valid timeout value.\n"
					        "Please specify an integer of value 0 or "
					        "greater.\n",
					        optarg);
					return false;
				}
				break;

			// --event or -e
			case 'e':
				// Get event mask from event string
				new_event = inotifytools_str_to_event(optarg);

				// If optarg was invalid, abort
				if ( new_event == -1 )
				{
					fprintf(stderr, "'%s' is not a valid event!  Run with the "
					                "'--help' option to see a list of "
					                "events.\n", optarg);
					return false;
				}

				// Add the new event to the event mask
				(*events) = ( (*events) | new_event );

				break;
				
			case 's':
				(*cb_script) =optarg;
				break;				


		}

		curr_opt = getopt_long(*argc, *argv, opt_string, long_opts, NULL);

	}

	if ( *monitor && *timeout != 0 ) {
		fprintf(stderr, "-m and -t cannot both be specified.\n");
		return false;
	}

	if ( *regex && *iregex ) {
		fprintf(stderr, "--exclude and --excludei cannot both be specified.\n");
		return false;
	}

	if ( *format && *csv ) {
		fprintf(stderr, "-c and --format cannot both be specified.\n");
		return false;
	}

	if ( !*format && *timefmt ) {
		fprintf(stderr, "--timefmt cannot be specified without --format.\n");
		return false;
	}

	if ( *format && strstr( *format, "%T" ) && !*timefmt ) {
		fprintf(stderr, "%%T is in --format string, but --timefmt was not "
		                "specified.\n");
		return false;
	}
	if ( !cb_script) {
		fprintf(stderr, "-s failed\n");
		return false;
	}
	
	(*argc) -= optind;
	*argv = &(*argv)[optind];

	// If ? returned, invalid option
	return (curr_opt != '?');
}


void print_help()
{
	printf("inotifywait %s\n", PACKAGE_VERSION);
	printf("Wait for a particular event on a file or set of files.\n");
	printf("Usage: inotifywait [ options ] file1 [ file2 ] [ file3 ] "
	       "[ ... ]\n");
	printf("Options:\n");
	printf("\t-h|--help     \tShow this help text.\n");
	printf("\t@<file>       \tExclude the specified file from being "
	       "watched.\n");
	printf("\t--exclude <pattern>\n"
	       "\t              \tExclude all events on files matching the\n"
	       "\t              \textended regular expression <pattern>.\n");
	printf("\t--excludei <pattern>\n"
	       "\t              \tLike --exclude but case insensitive.\n");
	printf("\t-m|--monitor  \tKeep listening for events forever.  Without\n"
	       "\t              \tthis option, inotifywait will exit after one\n"
	       "\t              \tevent is received.\n");
	printf("\t-r|--recursive\tWatch directories recursively.\n");
	printf("\t--fromfile <file>\n"
	       "\t              \tRead files to watch from <file> or `-' for "
	       "stdin.\n");
	printf("\t-q|--quiet    \tPrint less (only print events).\n");
	printf("\t-qq           \tPrint nothing (not even events).\n");
	printf("\t--format <fmt>\tPrint using a specified printf-like format\n"
	       "\t              \tstring; read the man page for more details.\n");
	printf("\t--timefmt <fmt>\tstrftime-compatible format string for use with\n"
	       "\t              \t%%T in --format string.\n");
	printf("\t-c|--csv      \tPrint events in CSV format.\n");
	printf("\t-t|--timeout <seconds>\n"
	       "\t              \tWhen listening for a single event, time out "
	       "after\n"
	       "\t              \twaiting for an event for <seconds> seconds.\n"
	       "\t              \tIf <seconds> is 0, inotifywait will never time "
	       "out.\n");
	printf("\t-e|--event <event1> [ -e|--event <event2> ... ]\n"
	       "\t\tListen for specific event(s).  If omitted, all events are \n"
	       "\t\tlistened for.\n\n");
	printf("Exit status:\n");
	printf("\t%d  -  An event you asked to watch for was received.\n",
	       EXIT_OK );
	printf("\t%d  -  An event you did not ask to watch for was received\n",
	       EXIT_ERROR);
	printf("\t      (usually delete_self or unmount), or some error "
	       "occurred.\n");
	printf("\t%d  -  The --timeout option was given and no events occurred\n",
	       EXIT_TIMEOUT);
	printf("\t      in the specified interval of time.\n\n");
	printf("Events:\n");
	print_event_descriptions();
}

